password-hash/src/ident.rs

```diff
@@ -0,0 +1,261 @@
+//! Algorithm or parameter identifier.
+//!
+//! Implements the following parts of the [PHC string format specification][1]:
+//!
+//! > The function symbolic name is a sequence of characters in: `[a-z0-9-]`
+//! > (lowercase letters, digits, and the minus sign). No other character is
+//! > allowed. Each function defines its own identifier (or identifiers in case
+//! > of a function family); identifiers should be explicit (human readable,
+//! > not a single digit), with a length of about 5 to 10 characters. An
+//! > identifier name MUST NOT exceed 32 characters in length.
+//! >
+//! > Each parameter name shall be a sequence of characters in: `[a-z0-9-]`
+//! > (lowercase letters, digits, and the minus sign). No other character is
+//! > allowed. Parameter names SHOULD be readable for a human user. A
+//! > parameter name MUST NOT exceed 32 characters in length.
+//!
+//! [1]: https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md
+
+use crate::errors::ParseError;
+use core::{
+    fmt,
+    ops::Deref,
+    str::{self, FromStr},
+};
+
+/// Maximum size of an identifier.
+const MAX_LENGTH: usize = 32;
+
+/// Algorithm or parameter identifier.
+///
+/// This type encompasses both the "function symbolic name" and "parameter name"
+/// use cases as described in the [PHC string format specification][1].
+///
+/// # Constraints
+/// - ASCII-encoded string consisting of the characters `[a-z0-9-]`
+///   (lowercase letters, digits, and the minus sign)
+/// - Minimum length: 1 ASCII character (i.e. 1-byte)
+/// - Maximum length: 32 ASCII characters (i.e. 32-bytes)
+///
+/// [1]: https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md
+#[derive(Copy, Clone, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct Ident {
+    /// Byte array containing an ASCII-encoded string.
+    bytes: [u8; MAX_LENGTH],
+
+    /// Length of the string in ASCII characters (i.e. bytes).
+    length: u8,
+}
```

Comment:
Note: this implementation contains some fairly repetitive `arrayvec`/`tinyvec`/`heapless::Vec`-like types which would probably benefit from using some sort of fix-sized `Vec`-like abstraction.

The decision not to use any of these is deliberate: avoiding any provides a baseline for experimentation as to which, if any, we should use.

With `min_const_generics` on the horizon it seemed best not to be too eager to pick, or to loop in `generic-array` quite yet.

The needs for such a type are quite minimal (these types are mostly immutable/allocate-once) with only a fixed size overhead, so a reasonable first step in a `min_const_generics` feature is to define a type internal to the library for representing them, and if that turns out to be too cumbersome, considering a dependency.

Reply:
Done in 739402d